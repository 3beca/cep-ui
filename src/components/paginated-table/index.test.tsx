import * as React from 'react';
import { render, fireEvent, within } from '@testing-library/react';
import PaginatedTable from './index';
import Table from '@material-ui/core/Table';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import TableCell from '@material-ui/core/TableCell';
import Typography from '@material-ui/core/Typography';
import TableBody from '@material-ui/core/TableBody';

const SampleTable: React.FC<{rows?: number}> = ({rows = 5}) => {
    return (
        <Table>
            <TableHead>
                <TableRow role='element header'>
                    <TableCell align='left'><Typography>Header</Typography></TableCell>
                </TableRow>
            </TableHead>
            <TableBody>
            {
                (new Array(rows)).map(
                    (row, index) => (
                        <TableRow role='element row' key={index}>
                            <TableCell align='right'><Typography>Data</Typography></TableCell>
                        </TableRow>
                    )
                )
            }
        </TableBody>
    </Table>
    );
};
test('PaginatedTable snapshot without data', () => {
    const { container } = render(<PaginatedTable isEmpty={true}/>);
    expect(container).toMatchSnapshot();
});

test('PaginatedTable snapshot with data and pagination', () => {
    const { container } = render(
        <PaginatedTable
            isLoading={false}
            page={2}
            size={5}
        >
            <SampleTable/>
        </PaginatedTable>

    );
    expect(container).toMatchSnapshot();
});

test('PaginatedTable snapshot reached last page', () => {
    const { container } = render(
        <PaginatedTable
            isLoading={false}
            isEmpty={true}
            page={2}
            size={5}
        >
            <SampleTable/>
        </PaginatedTable>

    );
    expect(container).toMatchSnapshot();
});

test('PaginatedTable snapshot while loading with previous data and pagination', () => {
    const { container } = render(
        <PaginatedTable
            isLoading={true}
            page={2}
            size={5}
        >
            <SampleTable/>
        </PaginatedTable>
    );
    expect(container).toMatchSnapshot();
});

test('PaginatedTable snapshot while loading without data', () => {
    const { container } = render(
        <PaginatedTable
            isLoading={true}
            page={1}
            size={5}
        />
    );
    expect(container).toMatchSnapshot();
});

test('PaginatedTable render without data should show the empty data message', async () => {
    // Render without data
    const {getByTitle, getByTestId, getByLabelText} = render(
        <PaginatedTable
            isLoading={false}
            isEmpty={true}
            page={1}
            size={5}
        />
    );
    const noDataView = getByTestId(/empty-view-row/i);
    const prevButton = getByTitle(/previous page/i) as HTMLButtonElement;
    const nextButton = getByTitle(/next page/i) as HTMLButtonElement;
    const rowsPerPage = getByLabelText(/^rows:$/i);

    expect(prevButton.disabled).toBe(true);
    expect(nextButton.disabled).toBe(true);
    expect(noDataView).toBeInTheDocument();
    expect(() => getByTestId(/loading-view-row/)).toThrowError();
    expect(rowsPerPage).toHaveTextContent('5');

});

test('PaginatedTable render while loading show the loading view', async () => {
    // Render without data
    const {getByTitle, getByLabelText, getByTestId} = render(
        <PaginatedTable
            isLoading={true}
            isEmpty={true}
            page={1}
            size={5}
        >
            <SampleTable/>
        </PaginatedTable>
    );

    const loadingView = getByTestId(/loading-view-row/i);
    const prevButton = getByTitle(/previous page/i) as HTMLButtonElement;
    const nextButton = getByTitle(/next page/i) as HTMLButtonElement;
    const rowsPerPage = getByLabelText(/^rows:$/i);

    expect(prevButton.disabled).toBe(true);
    expect(nextButton.disabled).toBe(true);
    expect(() => getByTestId(/empty-view-row/)).toThrowError();
    expect(loadingView).toBeInTheDocument();
    expect(rowsPerPage).toHaveTextContent('5');
});

test('PaginatedTable render items and can navigate by pages', async () => {
    const onChangePage = jest.fn();
    // Render in first page
    const {getByTitle, getByLabelText, rerender} = render(
        <PaginatedTable
            isLoading={false}
            page={1}
            size={5}
            hasNextPage={true}
            hasPrevPage={false}
            onChangePage={onChangePage}
        >
            <SampleTable/>
        </PaginatedTable>
    );
    const prevButton = getByTitle(/previous page/i) as HTMLButtonElement;
    const nextButton = getByTitle(/next page/i) as HTMLButtonElement;
    const rowsPerPage = getByLabelText(/^rows:$/i);

    expect(prevButton.disabled).toBe(true);
    expect(nextButton.disabled).toBe(false);
    expect(rowsPerPage).toHaveTextContent('5');

    // Simulate click on next page
    fireEvent.click(nextButton);
    expect(onChangePage).toHaveBeenCalledTimes(1);
    expect(onChangePage).toHaveBeenCalledWith(2);

    // Rerender in second page
    rerender(
        <PaginatedTable
            isLoading={false}
            page={2}
            size={5}
            hasNextPage={true}
            hasPrevPage={true}
            onChangePage={onChangePage}
        >
            <SampleTable/>
        </PaginatedTable>
    );

    expect(prevButton.disabled).toBe(false);
    expect(nextButton.disabled).toBe(false);
    expect(rowsPerPage).toHaveTextContent('5');

    // Simulate click on next page
    fireEvent.click(nextButton);
    expect(onChangePage).toHaveBeenCalledTimes(2);
    expect(onChangePage).toHaveBeenNthCalledWith(2, 3);

    // Rerender last page
    rerender(
        <PaginatedTable
            isLoading={false}
            page={3}
            size={5}
            hasNextPage={false}
            hasPrevPage={true}
            onChangePage={onChangePage}
        >
            <SampleTable/>
        </PaginatedTable>
    );

    expect(prevButton.disabled).toBe(false);
    expect(nextButton.disabled).toBe(true);
    expect(rowsPerPage).toHaveTextContent('5');

    // Simulate click on prev page
    fireEvent.click(prevButton);
    expect(onChangePage).toHaveBeenCalledTimes(3);
    expect(onChangePage).toHaveBeenNthCalledWith(3, 2);

    // Rerender in last page
    rerender(
        <PaginatedTable
            isLoading={false}
            page={2}
            size={5}
            hasNextPage={false}
            hasPrevPage={true}
            onChangePage={onChangePage}
        >
            <SampleTable/>
        </PaginatedTable>
    );

    expect(prevButton.disabled).toBe(false);
    expect(nextButton.disabled).toBe(true);
    expect(rowsPerPage).toHaveTextContent('5');
});

test('PaginatedTable render 10 elements when change pageSize to 20', async () => {
    // Render without data
    const onChangePageSize = jest.fn();
    const {rerender, getByTitle, getByLabelText, getByTestId} = render(
        <PaginatedTable
            isLoading={false}
            isEmpty={false}
            page={1}
            size={5}
            onChangePageSize={onChangePageSize}
        />
    );
    const prevButton = getByTitle(/Previous page/) as HTMLButtonElement;
    const nextButton = getByTitle(/Next page/) as HTMLButtonElement;
    const rowsPerPage = getByLabelText(/^rows:$/i);

    expect(prevButton.disabled).toBe(true);
    expect(nextButton.disabled).toBe(true);
    expect(() => getByTestId(/empty-view-row/)).toThrowError();
    expect(() => getByTestId(/loading-view-row/)).toThrowError();
    expect(rowsPerPage).toHaveTextContent('5');

    fireEvent.mouseDown(rowsPerPage);
    const bound = within(document.getElementById('menu-')!);
    const options = bound.getAllByRole(/option/);
    fireEvent.click(options[1]); // 0->5, 1->10, 2->20

    expect(onChangePageSize).toHaveBeenCalledTimes(1);
    expect(onChangePageSize).toHaveBeenCalledWith(10);

    rerender(
        <PaginatedTable
            isLoading={false}
            isEmpty={false}
            page={1}
            size={10}
            onChangePageSize={onChangePageSize}
        >
            <SampleTable rows={10}/>
        </PaginatedTable>
    );

    expect(prevButton.disabled).toBe(true);
    expect(nextButton.disabled).toBe(true);
    expect(() => getByTestId(/empty-view-row/)).toThrowError();
    expect(() => getByTestId(/loading-view-row/)).toThrowError();
    expect(rowsPerPage).toHaveTextContent('10');

    fireEvent.mouseDown(rowsPerPage);
    fireEvent.click(options[2]); // 0->5, 1->10, 2->20

    expect(onChangePageSize).toHaveBeenCalledTimes(2);
    expect(onChangePageSize).toHaveBeenNthCalledWith(2, 20);

    rerender(
        <PaginatedTable
            isLoading={false}
            isEmpty={false}
            page={1}
            size={20}
            onChangePageSize={onChangePageSize}
        >
            <SampleTable rows={20}/>
        </PaginatedTable>
    );

    expect(prevButton.disabled).toBe(true);
    expect(nextButton.disabled).toBe(true);
    expect(() => getByTestId(/empty-view-row/)).toThrowError();
    expect(() => getByTestId(/loading-view-row/)).toThrowError();
    expect(rowsPerPage).toHaveTextContent('20');
});
