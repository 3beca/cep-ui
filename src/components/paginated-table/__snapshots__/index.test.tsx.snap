// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PaginatedTable snapshot while loading with previous data and pagination 1`] = `
<div>
  <div
    class="MuiTableContainer-root"
  >
    <table
      class="MuiTable-root"
    >
      <thead
        class="MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head"
          role="element header"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft"
            scope="col"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Header
            </p>
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root"
      />
    </table>
  </div>
  <div
    class="makeStyles-loadingView-290"
    data-testid="loading-view-row"
  >
    <div
      class="MuiCircularProgress-root makeStyles-loadingSpinner-291 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      role="progressbar"
      style="width: 40px; height: 40px;"
    >
      <svg
        class="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </div>
  </div>
  <div
    class="MuiTablePagination-root"
    role="paginator"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
    >
      <div
        class="MuiTablePagination-spacer"
      />
      <p
        class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        Rows per page:
      </p>
      <div
        aria-label="pageSelector"
        class="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
      >
        <div
          aria-haspopup="listbox"
          aria-label="Rows per page:"
          class="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
          role="button"
          tabindex="0"
        >
          5
        </div>
        <input
          type="hidden"
          value="5"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <p
        class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        Page 2
      </p>
      <div
        class="MuiTablePagination-actions"
      >
        <button
          aria-label="Previous page"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          title="Previous page"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Next page"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          title="Next page"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`PaginatedTable snapshot while loading without data 1`] = `
<div>
  <div
    class="MuiTableContainer-root"
  />
  <div
    class="makeStyles-loadingView-374"
    data-testid="loading-view-row"
  >
    <div
      class="MuiCircularProgress-root makeStyles-loadingSpinner-375 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      role="progressbar"
      style="width: 40px; height: 40px;"
    >
      <svg
        class="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </div>
  </div>
  <div
    class="MuiTablePagination-root"
    role="paginator"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
    >
      <div
        class="MuiTablePagination-spacer"
      />
      <p
        class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        Rows per page:
      </p>
      <div
        aria-label="pageSelector"
        class="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
      >
        <div
          aria-haspopup="listbox"
          aria-label="Rows per page:"
          class="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
          role="button"
          tabindex="0"
        >
          5
        </div>
        <input
          type="hidden"
          value="5"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <p
        class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        Page 1
      </p>
      <div
        class="MuiTablePagination-actions"
      >
        <button
          aria-label="Previous page"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          title="Previous page"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Next page"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          title="Next page"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`PaginatedTable snapshot with data and pagination 1`] = `
<div>
  <div
    class="MuiTableContainer-root"
  >
    <table
      class="MuiTable-root"
    >
      <thead
        class="MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head"
          role="element header"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignLeft"
            scope="col"
          >
            <p
              class="MuiTypography-root MuiTypography-body1"
            >
              Header
            </p>
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root"
      />
    </table>
  </div>
  <div
    data-testid="empty-view-row"
  >
    <h4
      class="MuiTypography-root makeStyles-emptyView-170 MuiTypography-h4 MuiTypography-alignCenter"
    >
      No Elements created yet!
    </h4>
  </div>
  <div
    class="MuiTablePagination-root"
    role="paginator"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
    >
      <div
        class="MuiTablePagination-spacer"
      />
      <p
        class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        Rows per page:
      </p>
      <div
        aria-label="pageSelector"
        class="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
      >
        <div
          aria-haspopup="listbox"
          aria-label="Rows per page:"
          class="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
          role="button"
          tabindex="0"
        >
          5
        </div>
        <input
          type="hidden"
          value="5"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <p
        class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        Page 2
      </p>
      <div
        class="MuiTablePagination-actions"
      >
        <button
          aria-label="Previous page"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          title="Previous page"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Next page"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          title="Next page"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`PaginatedTable snapshot without data 1`] = `
<div>
  <div
    class="MuiTableContainer-root"
  />
  <div
    class="makeStyles-loadingView-2"
    data-testid="loading-view-row"
  >
    <div
      class="MuiCircularProgress-root makeStyles-loadingSpinner-3 MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      role="progressbar"
      style="width: 40px; height: 40px;"
    >
      <svg
        class="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          class="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx="44"
          cy="44"
          fill="none"
          r="20.2"
          stroke-width="3.6"
        />
      </svg>
    </div>
  </div>
  <div
    class="MuiTablePagination-root"
    role="paginator"
  >
    <div
      class="MuiToolbar-root MuiToolbar-regular MuiTablePagination-toolbar MuiToolbar-gutters"
    >
      <div
        class="MuiTablePagination-spacer"
      />
      <p
        class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        Rows per page:
      </p>
      <div
        aria-label="pageSelector"
        class="MuiInputBase-root MuiTablePagination-input MuiTablePagination-selectRoot"
      >
        <div
          aria-haspopup="listbox"
          aria-label="Rows per page:"
          class="MuiSelect-root MuiSelect-select MuiTablePagination-select MuiSelect-selectMenu MuiInputBase-input"
          role="button"
          tabindex="0"
        >
          10
        </div>
        <input
          type="hidden"
          value="10"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSelect-icon MuiTablePagination-selectIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M7 10l5 5 5-5z"
          />
        </svg>
      </div>
      <p
        class="MuiTypography-root MuiTablePagination-caption MuiTypography-body2 MuiTypography-colorInherit"
      >
        Page 1
      </p>
      <div
        class="MuiTablePagination-actions"
      >
        <button
          aria-label="Previous page"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          title="Previous page"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
              />
            </svg>
          </span>
        </button>
        <button
          aria-label="Next page"
          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit Mui-disabled Mui-disabled"
          disabled=""
          tabindex="-1"
          title="Next page"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
              />
            </svg>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
